packages/react/src/components/Toolbar/Toolbar.tsx

```diff
@@ -118,189 +97,348 @@ class Toolbar extends UIComponent<WithAsProp<ToolbarProps>, ToolbarState> {
-    currentItems: [],
-    stable: false,
-  }
+  overflowContainerRef = React.createRef<HTMLElement>()
+  overflowItemRef = React.createRef<HTMLElement>()
+  offsetMeasureRef = React.createRef<HTMLElement>()
+
+  // index of the last visible item in Toolbar, the rest goes to overflow menu
+  lastVisibleItemIndex: number
 
-  wrapperRef = React.createRef<HTMLElement>()
-  hiddenToolbarRef = React.createRef<HTMLElement>()
-  animationFrameId: number
+  renderStartTime: number // TODO: remove before merge
+  rtl: boolean
 
   handleItemOverrides = variables => predefinedProps => ({
     variables: mergeComponentVariables(variables, predefinedProps.variables),
   })
 
-  renderItems(items, variables) {
+  renderItems(items: ShorthandCollection<ToolbarItemProps, ToolbarItemShorthandKinds>, variables) {
     const itemOverridesFn = this.handleItemOverrides(variables)
-    return _.map(items, item => {
-      const kind = _.get(item, 'kind', 'item')
-
-      switch (kind) {
-        case 'divider':
-          return ToolbarDivider.create(item, { overrideProps: itemOverridesFn })
-        case 'group':
-          return ToolbarRadioGroup.create(item, { overrideProps: itemOverridesFn })
-        case 'toggle':
-          return ToolbarItem.create(item, {
-            defaultProps: { accessibility: toggleButtonBehavior },
-            overrideProps: itemOverridesFn,
-          })
-        case 'custom':
-          return ToolbarCustomItem.create(item, { overrideProps: itemOverridesFn })
-        default:
-          return ToolbarItem.create(item, { overrideProps: itemOverridesFn })
-      }
-    })
+    return _.map(
+      items,
+      (item: ShorthandValue<ToolbarItemProps & { kind?: ToolbarItemShorthandKinds }>) => {
+        const kind = _.get(item, 'kind', 'item')
+
+        switch (kind) {
+          case 'divider':
+            return ToolbarDivider.create(item, { overrideProps: itemOverridesFn })
+          case 'group':
+            return ToolbarRadioGroup.create(item, { overrideProps: itemOverridesFn })
+          case 'toggle':
+            return ToolbarItem.create(item, {
+              defaultProps: { accessibility: toggleButtonBehavior },
+              overrideProps: itemOverridesFn,
+            })
+          case 'custom':
+            return ToolbarCustomItem.create(item, { overrideProps: itemOverridesFn })
+          default:
+            return ToolbarItem.create(item, { overrideProps: itemOverridesFn })
+        }
+      },
+    )
   }
 
-  static getDerivedStateFromProps(props: ToolbarProps, state: ToolbarState) {
-    if (props.items === state.initialItems) {
-      return null
+  hide(el: HTMLElement) {
+    if (el.style.visibility === 'hidden') {
+      return
     }
 
-    return {
-      initialItems: props.items,
-      currentItems: props.items,
-      stableItems: state.stableItems,
-      stable: false,
+    el.style.visibility = 'hidden'
+    const wasFocusable = el.getAttribute('data-is-focusable')
```

Comment:
We have this as constant:

```suggestion
    const wasFocusable = el.getAttribute(IS_FOCUSABLE_ATTRIBUTE)
```